type checking torch/linalg

stmcgovern opened the pull request:
Fixes #149639

In an effort to improve the type checking coverage, added a stub file for the torch/linalg directory.

stmcgovern committed:
add type checking to linalg

stmcgovern committed:
run lintrunner

stmcgovern committed:
use torch.return_types and update some signatures in stub

stmcgovern committed:
update test with type information

Skylion007 commented on mypy.ini:
```diff
@@ -31,6 +31,7 @@ warn_unused_ignores = False
 
 files =
     torch,
+    torch/linalg,
```
This should be covered... why is this necessary... did you just need to do this invalidate the cache?

stmcgovern committed:
remove spurious dir and fix lint warnings

stmcgovern commented:
> @Skylion007 if this is ready to merge, would it be useful to add a stub to torch/fft and possibly other folders in the torch directory?

Hi, following up. Is the current PR ready to merge? cc @Skylion007

Skylion007 commented:
@pytorchbot merge

malfet reviewed (changes requested):
Having both `.py` and `.pyi` files are problematic as it will always result in out-of-sync issues

Why not add type annotations to `.py` file? And if they are for native functions, than perhaps they should be added to `_C/__init__.pyi.in` and then properly re-exported in `torch.linalg`

Also, would be nice to have a better test plan (automated) to avoid causing regression reported by @jeanschmidt

stmcgovern commented:
Hi @jeanschmidt, thanks for the information. Where are those tests that are failing? 

@malfet thanks for your questions and suggestion. The `__init__.pyi` takes precedence over `__init__.py` from tools/linter/adapters/mypy_linter.py, line 242. But it seems like adding individual `.pyi` stub files is not desirable. I will look into the alternative that you recommend. 
For testing and avoiding regressions, what would an automatic test plan look like? Adding more tests to the test suite?

malfet commented:
> For testing and avoiding regressions, what would an automatic test plan look like? Adding more tests to the test suite?

Yes, adding some test to `test_typing.py` would be nice
Also, if `torch.linalg` is consistent make sure it's checked by mypy by adding it to `mypy.ini`/`mypy-strict.init`

stmcgovern commented:
Thanks @malfet. I'll revisit this.

stmcgovern committed:
use torch/_C/__init__.pyi.in for linalg typing
tests pass - mypy needs fixing

stmcgovern committed:
remove the added pyi file

stmcgovern commented:
Hi @malfet, looking into this more, the Pylint E1102 error that the function is not callable applies to all the functions that use `_add_docstr` (torch.sparse, torch.special, torch.fft all have these errors) for their definition in `__init__.py` . To address this for linalg, I added an `if TYPE_CHECKING:` block in `linalg/__init__.py` importing the functions to be used for the static analysis tools and `else` using the `_add_docstr` for runtime. 

After removing `linalg/__init__.pyi` and placing the type signatures into `torch/_C/__init__.pyi.in`, I added `torch/_C/_linalg.pyi` to do the exporting of the module. I added `torch/linalg` to `mypy-strict.ini` (torch is already covered in `mypy.ini`). I added some tests to `test/test_typing.py`. 

When running `mypy --config=mypy.ini torch`
```
(base) [root@49023e5b8e19 pytorch]# mypy --config=mypy.ini torch
torch/distributions/constraints.py:637:16: error: "tuple[Tensor, Tensor]" has no attribute "info"  [attr-defined]
torch/functional.py:1840:35: error: Argument 3 to "linalg_vector_norm" has incompatible type "list[int | SymInt] | None"; expected "int | tuple[int, ...] | None"  [arg-type]
torch/functional.py:1844:35: error: Argument 3 to "linalg_vector_norm" has incompatible type "list[int | SymInt] | None"; expected "int | tuple[int, ...] | None"  [arg-type]
torch/functional.py:1852:59: error: Argument 3 to "linalg_matrix_norm" has incompatible type "list[int | SymInt]"; expected "tuple[int, int]"  [arg-type]
torch/functional.py:1855:31: error: Argument 3 to "linalg_matrix_norm" has incompatible type "list[int | SymInt]"; expected "tuple[int, int]"  [arg-type]
torch/functional.py:1861:60: error: Argument 3 to "linalg_vector_norm" has incompatible type "list[int | SymInt] | None"; expected "int | tuple[int, ...] | None"  [arg-type]
torch/functional.py:1864:32: error: Argument 3 to "linalg_vector_norm" has incompatible type "list[int | SymInt] | None"; expected "int | tuple[int, ...] | None"  [arg-type]
torch/nn/utils/spectral_norm.py:137:13: error: Argument 1 to "linalg_multi_dot" has incompatible type "list[Any]"; expected "tuple[Tensor, ...]"  [arg-type]
torch/_refs/__init__.py:3388:23: error: Argument 3 to "linalg_vector_norm" has incompatible type "list[int]"; expected "int | tuple[int, ...] | None"  [arg-type]
torch/_refs/__init__.py:3391:51: error: Argument 3 to "linalg_vector_norm" has incompatible type "list[int]"; expected "int | tuple[int, ...] | None"  [arg-type]
torch/_refs/__init__.py:5858:51: error: Argument 3 to "linalg_matrix_norm" has incompatible type "list[int] | tuple[int, ...]"; expected "tuple[int, int]"  [arg-type]
torch/_refs/__init__.py:5860:51: error: Argument 3 to "linalg_vector_norm" has incompatible type "list[int] | tuple[int, ...] | None"; expected "int | tuple[int, ...] | None"  [arg-type]
torch/masked/_ops.py:1582:25: error: Argument 2 to "linalg_vector_norm" has incompatible type "float | None"; expected "int | float"  [arg-type]
torch/_refs/nn/functional/__init__.py:1204:56: error: Argument "ord" to "linalg_vector_norm" has incompatible type "bool | int | float | complex"; expected "int | float"  [arg-type]
Found 14 errors in 6 files (checked 1948 source files)
```

It identifies some type errors. I hesitate to start making those changes in this same PR. What's the preferred way to go about addressing the uncovered type errors when enabling static type checking of torch/linalg?
cc @Skylion007 @jeanschmidt

Skylion007 commented:
Do a PR stack using ezyang's tool, forward fix the type errors then fix the code that checks the types (or even do it all at once honestly), it hopefully won't propagate too much.
